src/invariant/ConservationOfLumens.cpp

```diff
@@ -71,7 +79,54 @@ calculateDeltaBalance(LedgerEntry const* current, LedgerEntry const* previous)
     }
 #ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
     case CONTRACT_DATA:
-        break;
+    {
+        auto const& contractData = current ? current->data.contractData()
+                                           : previous->data.contractData();
+        if (contractData.contractID != lumenContractID ||
+            contractData.key.type() != SCV_VEC || !contractData.key.vec() ||
+            contractData.key.vec().size() == 0)
+        {
+            return 0;
+        }
+
+        // The balanceSymbol should be the first entry in the SCVec
+        if (!(contractData.key.vec()->at(0) == balanceSymbol))
+        {
+            return 0;
+        }
+
+        auto getAmount = [&amountSymbol](LedgerEntry const* entry) -> int64_t {
+            if (!entry)
+            {
+                return 0;
+            }
+
+            // The amount should be the first entry in the SCMap
+            auto const& val = entry->data.contractData().val;
+            if (val.type() == SCV_MAP && val.map() && val.map()->size() != 0)
+            {
+                auto const& amountEntry = val.map()->at(0);
+                if (amountEntry.key == amountSymbol)
+                {
+                    if (amountEntry.val.type() == SCV_I128)
+                    {
+                        auto lo = amountEntry.val.i128().lo;
+                        auto hi = amountEntry.val.i128().hi;
+                        if (lo > INT64_MAX || hi > 0)
+                        {
+                            // The amount isn't right, but it'll trigger the
+                            // invariant.
+                            return INT64_MAX;
+                        }
+                        return static_cast<int64_t>(lo);
```

Comment:
Do we allow negative amount? Since the type I128, I suppose it is possible unless we explicit assert it somewhere?
If the amount is negative, this will be incorrect. 

You'll probably need to do https://github.com/stellar/rs-stellar-xdr/blob/main/src/next/scval_conversions.rs#L215 here and cast the result to `int64_t`.

Reply:
Negative balances shouldn't be possible in the SAC, but maybe I shouldn't be making that assumption here?

Reply:
this would belong to the other invariant